libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -190,14 +199,44 @@ method initPubSub*(p: PubSub) {.base.} =
 
 method start*(p: PubSub) {.async, base.} =
   ## start pubsub
-  ## start long running/repeating procedures
   discard
 
 method stop*(p: PubSub) {.async, base.} =
   ## stopt pubsub
-  ## stop long running/repeating procedures
   discard
 
+method addValidator*(p: PubSub,
+                     topic: varargs[string],
+                     hook: ValidatorHandler) {.base.} =
+  for t in topic:
+    if t notin p.validators:
+      p.validators[t] = initHashSet[ValidatorHandler]()
+
+    trace "adding validator for topic", topicId = t
+    p.validators[t].incl(hook)
+
+method removeValidator*(p: PubSub,
+                        topic: varargs[string],
+                        hook: ValidatorHandler) {.base.} =
+  for t in topic:
+    if t in p.validators:
+      p.validators[t].excl(hook)
+
+method validate*(p: PubSub, message: Message): Future[bool] {.async, base.} =
+  var pending: seq[Future[bool]]
+  trace "about to validate message"
+  for topic in message.topicIDs:
+    trace "looking for validators on topic", topicID = topic,
+                                             registered = toSeq(p.validators.keys)
+    if topic in p.validators:
+      trace "running validators for topic", topicID = topic
+      # TODO: add timeout to validator
+      pending = pending.concat(p.validators[topic].mapIt(it(topic, message)))
```

Comment:
This code will allocate a new sequence on every concatenation. Use `pending.add` instead.